README.md
# Text Neon Widget (1.0.0)

[![pub package](https://img.shields.io/pub/v/text_neon_widget.svg)](https://pub.dev/packages/text_neon_widget)
[![pub points](https://img.shields.io/pub/points/text_neon_widget?color=2E8B57&label=pub%20points)](https://pub.dev/packages/text_neon_widget/score)




A Neon widget for Flutter: With this widget,
you can bring all your text to life with neon animations, and all elements are customizable,
allowing you to personalize them as you like.
For support, please follow me.

## Demo new

<div style="display: flex; flex-direction: row;">
  <img src="https://github.com/PuzzleTakX/text_neon_widget/blob/master/demo/1.jpg?raw=true" alt="image_demo" width="300" height="700">

  <img src="https://github.com/PuzzleTakX/text_neon_widget/blob/master/demo/2.jpg?raw=true" alt="second_image" width="300" height="700">
</div>


<img src="https://github.com/PuzzleTakX/text_neon_widget/blob/master/demo/4.gif?raw=true" alt="image_demo" width="300" height="700">

## Installation ☺

To use the NavigationView package, add the following dependency to your `pubspec.yaml`
file:

[![pub package](https://img.shields.io/pub/v/text_neon_widget.svg)](https://pub.dev/packages/text_neon_widget)
```yaml
dependencies:
text_neon_widget : ^1.0.5
```
Then, run `flutter pub get` to fetch the package.

## Usage

Import the package in your Dart file:

```dart
import 'package:text_neon_widget/text_neon_widget.dart';
```

Create a `PTTextNeon` widget and provide the necessary parameters:

```dart
     Row(
    mainAxisAlignment: MainAxisAlignment.center,
    children: [
            PTTextNeon(text: 'F l u t t e r ',color: Colors.lightBlue,
            font: "five",shine: true,fontSize: 40,
            strokeWidthTextHigh: 3,blurRadius: 25,
            strokeWidthTextLow: 1,backgroundColor: Colors.black,),
            
            PTTextNeon(text: ' & ',color: Colors.deepPurple,
            font: "five",shine: true,fontSize: 25,
            strokeWidthTextHigh: 3,blurRadius: 25,
            strokeWidthTextLow: 1,backgroundColor: Colors.black,),
            
            PTTextNeon(text: ' D a r t',color: Colors.teal,
            shine: true,fontSize: 40,
            font: "five",strokeWidthTextHigh: 3,blurRadius: 20,
            strokeWidthTextLow: 1,backgroundColor: Colors.black,),
        ],
    )
```

In the above code, the `text` parameter specifies the text to be displayed. The `color` parameter sets the color of the neon glow effect. The `font` parameter allows you to specify the font for the text.

You can enable or disable the shine effect using the `shine` parameter. The `fontSize` parameter controls the size of the text. The `strokeWidthTextHigh` and `strokeWidthTextLow` parameters define the stroke width of the text in the high and low intensity areas respectively.

The `blurRadius` parameter sets the blur radius of the neon glow effect. The `backgroundColor` parameter sets the background color of the text.

Feel free to experiment with different values for the parameters to achieve the desired neon text effect.

```dart
     PTTextNeon PTTextNeon({
      required String text,
      required MaterialColor color,
      String? font,
      double? fontSize,
      NeonLevel? levelNeon,
      Color? backgroundColor,
      int? highAlpha,
      int? lowAlpha,
      Duration? animatedChangeDuration,
      double? strokeWidthTextLow,
      double? strokeWidthTextHigh,
      double? blurRadius,
      bool? shine,
      Duration? shineDuration,
      TextStyle? textStyle, }) 
```

## Example

For a complete example of using the NavigationView package, refer to
the [example](https://github.com/PuzzleTakX/text_neon_widget/tree/master/example) provided.

## License

This package is released under the MIT License. See the [LICENSE](https://github.com/PuzzleTakX/text_neon_widget/blob/master/LICENSE)
file for more details.

## Credits

NavigationView is developed and maintained by [puzzleTak](https://github.com/PuzzleTakX).

CHANGELOG.md
# PuzzleTak Neon Widget  (1.0.0)

A Neon widget for Flutter: With this widget,
you can bring all your text to life with neon animations, and all elements are customizable,
allowing you to personalize them as you like.
For support, please follow me.

## Demo new

<div style="display: flex; flex-direction: row;">
  <img src="https://github.com/PuzzleTakX/puzzletak_neon_widget/blob/master/demo/1.jpg?raw=true" alt="image_demo" width="300" height="700">

  <img src="https://github.com/PuzzleTakX/puzzletak_neon_widget/blob/master/demo/2.jpg?raw=true" alt="second_image" width="300" height="700">
</div>


<img src="https://github.com/PuzzleTakX/puzzletak_neon_widget/blob/master/demo/4.gif?raw=true" alt="image_demo" width="300" height="700">

## Installation ☺

To use the NavigationView package, add the following dependency to your `pubspec.yaml`
file:


```yaml
dependencies:
puzzletak_neon_widget : ^1.0.0
```
Then, run `flutter pub get` to fetch the package.

## Usage

Import the package in your Dart file:

```dart
import 'package:puzzletak_neon_widget/puzzletak_neon_widget.dart';
```

Create a `PTTextNeon` widget and provide the necessary parameters:

```dart
     Row(
    mainAxisAlignment: MainAxisAlignment.center,
    children: [
            PTTextNeon(text: 'F l u t t e r ',color: Colors.lightBlue,
            font: "five",shine: true,fontSize: 40,
            strokeWidthTextHigh: 3,blurRadius: 25,
            strokeWidthTextLow: 1,backgroundColor: Colors.black,),
            
            PTTextNeon(text: ' & ',color: Colors.deepPurple,
            font: "five",shine: true,fontSize: 25,
            strokeWidthTextHigh: 3,blurRadius: 25,
            strokeWidthTextLow: 1,backgroundColor: Colors.black,),
            
            PTTextNeon(text: ' D a r t',color: Colors.teal,
            shine: true,fontSize: 40,
            font: "five",strokeWidthTextHigh: 3,blurRadius: 20,
            strokeWidthTextLow: 1,backgroundColor: Colors.black,),
        ],
    )
```

In the above code, the `text` parameter specifies the text to be displayed. The `color` parameter sets the color of the neon glow effect. The `font` parameter allows you to specify the font for the text.

You can enable or disable the shine effect using the `shine` parameter. The `fontSize` parameter controls the size of the text. The `strokeWidthTextHigh` and `strokeWidthTextLow` parameters define the stroke width of the text in the high and low intensity areas respectively.

The `blurRadius` parameter sets the blur radius of the neon glow effect. The `backgroundColor` parameter sets the background color of the text.

Feel free to experiment with different values for the parameters to achieve the desired neon text effect.

```dart
     PTTextNeon PTTextNeon({
      required String text,
      required MaterialColor color,
      String? font,
      double? fontSize,
      NeonLevel? levelNeon,
      Color? backgroundColor,
      int? highAlpha,
      int? lowAlpha,
      Duration? animatedChangeDuration,
      double? strokeWidthTextLow,
      double? strokeWidthTextHigh,
      double? blurRadius,
      bool? shine,
      Duration? shineDuration,
      TextStyle? textStyle, }) 
```

## Example

For a complete example of using the NavigationView package, refer to
the [example](https://github.com/PuzzleTakX/puzzletak_neon_widget/tree/master/example) provided.

## License

This package is released under the MIT License. See the [LICENSE](https://github.com/PuzzleTakX/puzzletak_neon_widget/blob/master/LICENSE)
file for more details.

## Credits

NavigationView is developed and maintained by [puzzleTak](https://github.com/PuzzleTakX).

test/text_neon_widget.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:text_neon_widget/text_neon_widget.dart';



void main() {
  test('adds one to input values', () {
    return const PTTextNeon(text: "PuzzleTak", color: Colors.blue);
  });
}


test/puzzletak_neon_widget_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:puzzletak_neon_widget/puzzletak_neon_widget.dart';

void main() {
  test('adds one to input values', () {
    return const PTTextNeon(text: "PuzzleTak", color: Colors.blue);
  });
}


lib/puzzletak_neon_widget.dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';

enum NeonLevel { lowLevelNeon, highLevelNeon }
class PTTextNeon extends StatefulWidget {
  final String text;
  final MaterialColor color;
  final String? font;
  final double? fontSize;
  final NeonLevel? levelNeon;
  final double? blurRadius;
  final int? lowAlpha;
  final int? highAlpha;
  final bool? shine;
  final double? strokeWidthTextLow;
  final double? strokeWidthTextHigh;
  final Color? backgroundColor;
  final Duration? shineDuration;
  final Duration? animatedChangeDuration;
  final TextStyle? textStyle;

  const PTTextNeon(
      {Key? key,required this.text,required this.color,this.font, this.fontSize, this.levelNeon,
        this.backgroundColor,
        this.highAlpha,
        this.lowAlpha,
        this.animatedChangeDuration,
        this.strokeWidthTextLow,
        this.strokeWidthTextHigh,
        this.blurRadius, this.shine,this.shineDuration, this.textStyle,})
      : super(key: key);

  @override
  // ignore: library_private_types_in_public_api
  _PTTextNeonState createState() => _PTTextNeonState();
}

class _PTTextNeonState extends State<PTTextNeon> with TickerProviderStateMixin {
  AnimationController? _shadowController;

  String get text => widget.text;
  NeonLevel levelNeon = NeonLevel.highLevelNeon;
  MaterialColor get color => widget.color;
  TextStyle? get textStyle => widget.textStyle;
  late Timer timer ;
  bool hasTimer = false;

  @override
  void initState() {
    if(widget.levelNeon != null) levelNeon = widget.levelNeon!;
    _shadowController = AnimationController(
      duration: (widget.shineDuration == null) ? const Duration(milliseconds: 800) : widget.shineDuration,
      vsync: this,
    );
    if (_shine()) {
      _shadowController!.repeat(min: 0.2, max: 1, reverse: true);
    }
    hasTimer = true;
    initTimer();
    super.initState();
  }


  void initTimer(){
    timer = Timer.periodic((widget.animatedChangeDuration == null) ? const Duration(milliseconds: 2500) : widget.animatedChangeDuration!, (timer) {
      if(hasTimer){
        int r = Random().nextInt(2);
        if(r == 1){
          levelNeon = NeonLevel.lowLevelNeon;
        }else{
          levelNeon = NeonLevel.highLevelNeon;
        }
      }else{
        if(timer.isActive) timer.cancel();
      }
      setState(() {});
    });
  }

  @override
  void didUpdateWidget(PTTextNeon oldWidget) {
    super.didUpdateWidget(oldWidget);

  }

  @override
  void dispose() {
    timer.cancel();
    _shadowController?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
        child: _shine()
            ? AnimatedBuilder(
          animation: _shadowController!,
          builder: (context, child) {
            var radius = _shadowController!.value * _blurRadius() * 1.5;
            return Stack(
              children: <Widget>[
                Text(text,
                    style: textStyle != null
                        ? textStyle!.copyWith(
                        fontFamily: (_font()),
                        foreground: Paint()
                          ..style = PaintingStyle.stroke
                          ..invertColors = false
                          ..strokeWidth = _strokeWidthTextHigh()
                          ..color = _getLevelColor(levelNeon)!,
                        fontSize: _fontSize(),
                        shadows: _getShadows(levelNeon, radius))
                        : TextStyle(
                        fontFamily: (_font()),
                        foreground: Paint()
                          ..style = PaintingStyle.stroke
                          ..invertColors = false
                          ..strokeWidth = _strokeWidthTextHigh()
                          ..color = _getLevelColor(levelNeon)!,
                        fontSize: _fontSize(),
                        shadows: _getShadows(levelNeon, radius))),
                (widget.backgroundColor == null) ? const SizedBox(width: 0,height: 0,): Text(text,
                    style: textStyle != null
                        ? textStyle!.copyWith(
                        fontFamily: (_font()),
                        color: levelNeon == NeonLevel.highLevelNeon  ? _backgroundColor().withAlpha(_lowAlpha()) : _backgroundColor().withAlpha(_highAlpha()),
                        fontSize: _fontSize(),
                        shadows: _getShadows(levelNeon, radius))
                        : TextStyle(
                        fontFamily: (_font()),
                        color: levelNeon == NeonLevel.highLevelNeon  ? _backgroundColor().withAlpha(_lowAlpha()) : _backgroundColor().withAlpha(_highAlpha()),
                        fontSize: _fontSize(),
                        shadows: _getShadows(levelNeon, radius)))
              ],
            );
          },
        )
            : Stack(
          children: <Widget>[
            Text(text,
                style: TextStyle(
                    fontFamily: (_font()),

                    fontSize: _fontSize(),
                    foreground: Paint()
                      ..style = PaintingStyle.stroke
                      ..strokeWidth = _strokeWidthTextLow()
                      ..color = _getLevelColor(levelNeon)!,
                    shadows: _getShadows(levelNeon, _blurRadius()))),
                (widget.backgroundColor == null) ? const SizedBox(width: 0,height: 0,): Text(text,
                style: textStyle != null
                    ? textStyle!.copyWith()
                    : TextStyle(
                    fontFamily: (_font()),
                    color: levelNeon == NeonLevel.highLevelNeon  ? _backgroundColor().withAlpha(_lowAlpha()) : _backgroundColor().withAlpha(_highAlpha()),
                    fontSize: _fontSize(),
                    shadows: _getShadows(levelNeon, _blurRadius())))
          ],
        ));
  }



  /// check data add
  Color _backgroundColor(){
    return (widget.backgroundColor == null) ? Colors.black : widget.backgroundColor!;
  }
  int _lowAlpha(){
    return (widget.lowAlpha == null) ? 150 : widget.lowAlpha!;
  }
  int _highAlpha(){
    return (widget.highAlpha == null) ? 150 : widget.highAlpha!;
  }
  double _strokeWidthTextLow(){
    return (widget.strokeWidthTextLow == null) ? 1 : widget.strokeWidthTextLow!;
  }

  double _strokeWidthTextHigh(){
    return (widget.strokeWidthTextHigh == null) ? 3 : widget.strokeWidthTextHigh!;
  }
  String _font(){
    return (widget.font == null) ? "" : widget.font!;
  }
  double _fontSize(){
    return (widget.fontSize == null) ? 30 : widget.fontSize!;
  }
  double _blurRadius(){
    return (widget.blurRadius == null) ? 30 : widget.blurRadius!;
  }
  bool _shine(){
    return (widget.shine == null) ? true : widget.shine!;
  }





  List<Shadow> _getShadows(NeonLevel? levelNeon, double radius) {
    if (levelNeon == NeonLevel.lowLevelNeon) {
      return [
        Shadow(color: color[500]!, blurRadius: radius / 6),
      ];
    } else {
      return [
        Shadow(color: color[300]!, blurRadius: radius / 2),
        Shadow(color: color[400]!, blurRadius: radius),
        Shadow(color: color[500]!, blurRadius: radius * 3),
      ];
    }
  }

  Color? _getLevelColor(NeonLevel? levelNeon) {
    return levelNeon == NeonLevel.lowLevelNeon ? color[400]!.withAlpha(50) : color[100];
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:text_neon_widget/text_neon_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'PuzzleTak Neon Widget',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<ItemColor> list = [
    ItemColor(text: "P",color: Colors.blue),
    ItemColor(text: "U",color: Colors.amber),
    ItemColor(text: "Z",color: Colors.red),
    ItemColor(text: "Z",color: Colors.green),
    ItemColor(text: "L",color: Colors.indigo),
    ItemColor(text: "E",color: Colors.pink),
    ItemColor(text: "-",color: Colors.deepOrange),
    ItemColor(text: "T",color: Colors.lime),
    ItemColor(text: "A",color: Colors.teal),
    ItemColor(text: "K",color: Colors.lightGreen),
  ];

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        title: const Text('PuzzleTak Neon Widget'),
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
      ),
      backgroundColor: Colors.black,
      body: SingleChildScrollView(
       child: Column(
         children: [
           const SizedBox(height: 50,),
           _simpleOne(),
           const SizedBox(height: 30,),
           _simpleTow(),
           const SizedBox(height: 30,),
           _simpleTree(),
           const SizedBox(height: 30,),
           _simpleFour(),

         ],
       ),
      ),
    );
  }

  Widget _simpleOne(){
    return const Row(
      textDirection: TextDirection.rtl,
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        PTTextNeon(text: 'بهمن',color: Colors.blue,
          font: "one",
          shine: true,
          fontSize: 30,
          strokeWidthTextHigh: 3,
          blurRadius: 30,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,

        ),
        PTTextNeon(text: '  تیموری  ',color: Colors.red,
          shine: true,
          font: "one",
          fontSize: 30,
          strokeWidthTextHigh: 3,
          blurRadius: 30,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,

        ),
        PTTextNeon(text: 'نژاد',color: Colors.green,
          shine: true,
          fontSize: 30,
          font: "one",
          strokeWidthTextHigh: 3,
          blurRadius: 30,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,



        ),
      ],
    );
  }
  Widget _simpleTow(){
    return const Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        PTTextNeon(text: 'N e o n ',color: Colors.amber,
          font: "four",
          shine: true,
          fontSize: 40,
          strokeWidthTextHigh: 3,
          blurRadius: 25,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,

        ),
        PTTextNeon(text: 'W i d g e t',color: Colors.pink,
          shine: true,
          fontSize: 40,
          font: "four",
          strokeWidthTextHigh: 3,
          blurRadius: 20,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,

        ),
      ],
    );
  }
  Widget _simpleTree(){
    return const Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        PTTextNeon(text: 'F l u t t e r ',color: Colors.lightBlue,
          font: "five",
          shine: true,
          fontSize: 40,
          strokeWidthTextHigh: 3,
          blurRadius: 25,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,

        ),
        PTTextNeon(text: ' & ',color: Colors.deepPurple,
          font: "five",
          shine: true,
          fontSize: 25,
          strokeWidthTextHigh: 3,
          blurRadius: 25,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,

        ),
        PTTextNeon(text: ' D a r t',color: Colors.teal,
          shine: true,
          fontSize: 40,
          font: "five",
          strokeWidthTextHigh: 3,
          blurRadius: 20,
          strokeWidthTextLow: 1,
          backgroundColor: Colors.black,

        ),
      ],
    );
  }
  Widget _simpleFour(){
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: list.map((e) => PTTextNeon(text: " ${e.text!} ",color: e.color!,
        font: "tree",
        shine: true,
        fontSize: 35,
        strokeWidthTextHigh: 3,
        blurRadius: 25,
        strokeWidthTextLow: 1,
        backgroundColor: Colors.black,

      ),).toList(),
    );
  }




}


class ItemColor{
   String? text;
   MaterialColor? color;

   ItemColor({this.color,this.text});

}
